### DISCUSSION TITLE
Custom Auth Provider Integration

### DISCUSSION THREAD

**Discussion by Kay2dan** [nextauthjs/next-auth#2291] [Help]
Hi, I am trying to connect next-auth to our custom OAuth-2.0 implementation.

The `get access token` test via Postman works successfully and I am able to retrieve a jwk token. However, next-auth redirects back without the jwtoken.

Code:

```javascript
export default NextAuth({
  providers: [
    {
      id: "quot-test",
      name: "quot-test",
      type: "oauth",
      version: "2.0",
      scope: "openid email",
      authorizationParams: {
        response_type: "code",
      },
      authorizationUrl: "http://localhost:9000/oauth2/authorize",
      accessTokenUrl: "http://localhost:9000/oauth2/token",
      requestTokenUrl: "http://localhost:9000/oauth2/jwks",
      profile: (profile, tokens) => {
        console.log(tokens);
        return tokens;
      },
      clientId: process.env.QUOT_ID,
      clientSecret: process.env.QUOT_SECRET,
    },
  ],
  debug: true,
  session: {
    jwt: true,
    maxAge: 60 * 15,
  },
});
```

On client, the session is being returned as null. I am a bit confused whether we have to pass custom callback functions for session & jwt for our custom provider?

Thank you

Edit: Added more detail

**Comment by balazsorban44** [comment]
yes, you will need to utilize the jwt and session callbacks here.

**Reply by Kay2dan** [reply]
Thank you for getting back to me. When I add the following, the logs are not printed when I go through the login process.

```
  callbacks: {
    jwt: async (token, user, account, profile, isNewUser) => {
      console.log(token);
      console.log(user);
      console.log(account);
      ...
```

Secondly,  are there any example of implementation of these callbacks for custom provider?

**Reply by balazsorban44** [reply]
https://next-auth.js.org/configuration/callbacks#jwt-callback

https://next-auth.js.org/configuration/callbacks#session-callback

nothing special about these in terms of custom providers.

Regarding your issue, I would check your config. are you using an adapter/database? jwt is not invoked then. do you have a reproduction/public repo somewhere?

**Reply by Kay2dan** [reply]
Thank you so much for your help. I dont have permission to upload the ref to private authentication service. 

More info:

- I am not using a db and expect a jwt.
- The jwt is unencrypted.
- My .env.local file has only three keys, along with their values defined:
  - NEXTAUTH_URL, QUOT_ID & QUOT_SECRECT.

However, my new `[...nextauth].js` looks as follows:

```javascript
import NextAuth from "next-auth";

export default NextAuth({
  providers: [
    {
      id: "quot-test",
      name: "quot-test",
      type: "oauth",
      version: "2.0",
      scope: "openid email",
      state: false,
      protection: "none",
      authorizationParams: {
        response_type: "code",
      },
      idToken: true,
      authorizationUrl: "http://localhost:9000/oauth2/authorize",
      accessTokenUrl: "http://localhost:9000/oauth2/token",
      requestTokenUrl: "http://localhost:9000/oauth2/jwks",
      profile: (profile, tokens) => tokens,
      clientId: process.env.QUOT_ID,
      clientSecret: process.env.QUOT_SECRET,
    },
  ],
  debug: true,
  session: {
    jwt: true,
    maxAge: 60 * 15,
  },
  callbacks: {
    signIn: async (user, account, profile) => true,
    jwt: async (token, user, account, profile, isNewUser) => {
      console.log(token);
      console.log(user);
      console.log(account);
      console.log(profile);
      console.log(isNewUser);
      return Promise.resolve(token);
    },
    session: async (session, token) => {
      console.log(session);
      console.log(token);
      return session;
    },
  },
});
```
As none of the consoles are being printed, that is my first problem.

Secondly, I am confused by this comment : https://github.com/nextauthjs/next-auth/issues/620#issuecomment-683878620

What are next-auth routes that can access the token? As we dont have encrypted tokens, do we need to create a separate secret (in addition to clientSecret) so we can access the token? And do I need to create a getToken function to be able to access the token?

**Reply by Kay2dan** [reply]
I have detailed my steps in [this stackoverflow](https://stackoverflow.com/questions/68225050/next-auth-custom-oauth-2-0-provider-integration) post if that would help. Would be most grateful for guidance with this issue.

**Reply by balazsorban44** [reply]
You probably don't want to return the whole tokens object in profile rather a subset of it. I'm fairly sure you will need to return an `id` in any case.

we will need more info/logs why jwt is not called.

getToken is imported from `next-auth/jwt`
https://next-auth.js.org/tutorials/securing-pages-and-api-routes#using-gettoken

for encryption, see https://next-auth.js.org/configuration/options#jwt

**Reply by Kay2dan** [reply]
Thank you @balazsorban44 . 

Edit:

Here is the test repo with the settings configured as per my understanding. Do note, inputting the above provider settings in postman retrieves the correct `jwt` token, so clearly I am missing something in my implementation.

https://github.com/Kay2dan/new_auth.git

### The .env.local file has 4 keys:

NEXTAUTH_URL=http://localhost:3000/

QUOT-ID=quot-test

QUOT-SECRET=[key]

SECRET=[random string]

I hope this helps.
